source/fuzz/fuzzer_pass_add_composite_inserts.cpp

```diff
@@ -0,0 +1,207 @@
+          auto current_node_type_inst =
+              GetIRContext()->get_def_use_mgr()->GetDef(current_node_type_id);
+          uint32_t num_of_components = fuzzerutil::GetBoundForCompositeIndex(
+              *current_node_type_inst, GetIRContext());
+
+          // If the composite is empty, then end the iteration.
+          if (num_of_components == 0) {
+            break;
+          }
+          uint32_t one_selected_index =
+              GetFuzzerContext()->GetRandomIndexForCompositeInsert(
+                  num_of_components);
+
+          // Construct a final index by appending the current index.
+          path_to_replaced.push_back(one_selected_index);
+          current_node_type_id = fuzzerutil::WalkOneCompositeTypeIndex(
+              GetIRContext(), current_node_type_id, one_selected_index);
+
+          // If the component is not a composite then end the iteration.
+          if (!fuzzerutil::IsCompositeType(
+                  GetIRContext()->get_type_mgr()->GetType(
+                      current_node_type_id))) {
+            break;
+          }
+
+          // If the component is a composite, but we decide not to go deeper,
+          // then end the iteration.
+          if (!GetFuzzerContext()->ChoosePercentage(
+                  GetFuzzerContext()
+                      ->GetChanceOfGoingDeeperToInsertInComposite())) {
+            break;
+          }
+        }
+
+        // Look for available objects that have the type id
+        // |current_node_type_id| and can be inserted.
+        std::vector<opt::Instruction*> available_objects =
+            FindAvailableInstructions(
+                function, block, instruction_iterator,
+                [instruction_descriptor, current_node_type_id](
+                    opt::IRContext* /*unused*/,
+                    opt::Instruction* instruction) -> bool {
+                  if (instruction->result_id() == 0 ||
+                      instruction->type_id() == 0) {
+                    return false;
+                  }
+                  if (instruction->type_id() != current_node_type_id) {
+                    return false;
+                  }
+                  return true;
+                });
+
+        // If there are no objects of the specific type available, create a zero
+        // constant of this type, which is not a pointer.
+        // TODO: (https://github.com/KhronosGroup/SPIRV-Tools/issues/3658)
+        //       Structs can have components of pointer type.
+        //       FindOrCreateZeroConstant cannot be called on a pointer. We
+        //       ignore pointers for now. Consider adding support for pointer
+        //       types.
+        uint32_t available_object_id;
+        if (available_objects.empty()) {
+          auto current_node_type =
+              GetIRContext()->get_type_mgr()->GetType(current_node_type_id);
+          if (current_node_type->kind() == opt::analysis::Type::kPointer) {
+            return;
+          }
```

Comment:
You only ignore pointer types in the case where you try to create a zero constant. In the other case (an existing, available id), the id might have a pointer type. However, in IsApplicable, you return false if the id has a pointer type. Thus, this fuzzer pass could fail if it finds a struct with a member with pointer type and an existing id to insert of the required type. I suggest that you (for now) ignore any composites with pointer types (like you do for OpRuntimeArray). Or you could remove the check from IsApplicable, but it might be hard to write a test case for a struct with a pointer type (and I doubt we will see this often).